### ISSUE TITLE
Runtime validation of configuration parameters

### ISSUE THREAD

**Issue by kaikreuzer** [eclipse/smarthome#587]
migrated from Bugzilla [#465428](https://bugs.eclipse.org/bugs/show_bug.cgi?id=465428)
status NEW severity *enhancement* in component *Core* for *---*
Reported in version *unspecified* on platform *All*
Assigned to: Project Inbox

On 2015-04-24 09:53:11 -0400, Chris Jackson wrote:

> Currently, when saving a parameter, ESH is not checking the type that is in the thing XML file (at least this is my working assumption from looking at what is happening). The problem is that if we send a parameter "6", then ESH saves this as a string, and then restores it as a string when the getConfig function is called. However, if the binding thinks this is going to be an integer, and assumes this (as per our previous discussion) then it throws an exception.  The way to solve this is for the rest interface to write this as 6, not "6".
> 
> I know that you could (and probably will) consider this a bug in the UI and I agree that the UI should send back the data correctly. However given the consequences that it will cause the binding to crash if it's wrong, I think this should be checked, and converted where possible, in ESH.
> 
> The other option is for the binding to perform type checking to see what is returned from getConfig, but this seems wrong to me. The binding has stated it wants an Integer, so it should get one (or at least a BigInteger as per our previous discussion).

On 2015-04-25 08:17:09 -0400, Dennis Nobel wrote:

> Hi Chris,
> 
> first I think that the application should send correct values. In the Paper UI IÂ´m converting the values at the client. If the binding expects an integer or boolean, the Paper UI sends the native type as part of the REST call:
> 
> 'configuration': {
>    'stringParameter': 'value',
>    'decimalParameter': 20,
>    'booleanParameter': true
> }
> 
> But I agree that we should have type checking as part of the framework. We planned to do so, but as usual we did not find the time to implement it. 
> 
> > The other option is for the binding to perform type checking to see what is returned from getConfig, but this seems wrong to me. The binding has stated it wants an Integer, so it should get one (or at least a BigInteger as per our previous discussion).
> 
> Yes, converting inside the binding makes no sense, as (ideally) the binding should not get wrong values if we add type checking and config validation.
> 
> Regards
> 
> Dennis

On 2015-04-25 08:22:23 -0400, Dennis Nobel wrote:

> If it is ok for you, I will change the title of the bug to "Runtime validation of configuration parameters" and put it to the list of feature requests.

On 2015-06-27 08:06:18 -0400, Chris Jackson wrote:

> Dennis,
> Just another example of this issue. I have the following parameter defined -:
>             <parameter name="channel" type="integer" required="true">
>                 <label>Channel</label>
>                 <description>Specifies the channel the sensor is using</description>
>                 <options>
>                 	<option value="1">Channel 1</option>
>                 	<option value="2">Channel 2</option>
>                 	<option value="3">Channel 3</option>
>                 	<option value="4">Channel 4</option>
>                 	<option value="5">Channel 5</option>
>                 	<option value="6">Channel 6</option>
>                 	<option value="7">Channel 7</option>
>                 	<option value="8">Channel 8</option>
>                 </options>
>             </parameter>
> 
> This is saved by paper UI as a string and I then get an exception in my binding. I think this might be a bug in paper ui though since it works ok in HABmin.
> 
> I missed the comment when you said you would change the name to 'validation' -  to me it should be conversion as well as validation (ie conversion where possible, and validation where not)  Are you happy with that definition? IF so, I might have a look at fixing this.

On 2015-06-27 08:19:25 -0400, Chris Jackson wrote:

> Actually, what I described above might be a bug specific to PaperUI when creating a new thing. If I edit the parameters, then it is correctly saved as a BigInt, but when it was first created, it was saved as a String.

On 2015-06-28 12:40:23 -0400, Dennis Nobel wrote:

> Yes, itÂ´s a bug in the Paper UI (see https://github.com/openhab/smarthome/issues/15). As commented before I would only expect a runtime validation, not a conversation. UIs should send data in the correct format.

On 2015-08-01 04:55:53 -0400, Chris Jackson wrote:

> When adding the validation, we need to ensure that arrays are supported (for the multiple selection option)

On 2015-08-03 09:43:39 -0400, Dennis Nobel wrote:

> Maybe I missed the forum discussion or the PR, but what is the format of an array value inside the configuration? Values just separated by "," or ";"? Or a real JSON Object like ["a", "b"]?

**Comment by cdjackson** [comment]
It looks like if the UI doesn't set a value when saving the configuration (ie the DTO doesn't have an entry for this parameter), then we end up with a blank string - probably this is how the DTO gets processed through Jetty. This blank string then can't be parsed in the thing handler if it's expecting a BigDecimal.

The big problem here is with the way this is coded, it seems the only way out is to force delete the thing, then add it back since once the incorrect data gets into the config, the thing handler throws an exception, and we can't update the config - so we're stuck.

Implementing some sort of validation here would at least protect the system from rogue UIs.

**Comment by kaikreuzer** [comment]
Does this happen on an initial Thing creation or upon a configuration change? As you say "the thing handler throws an exception", I assume that it is in the initialize().

> we can't update the config

Why can't we? The updated configuration should be received by the ThingHandler and thus it should have a chance to recover, I would think?

Note that the `BaseThingHandler` also already does a [validation of configuration updates](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L115) against the config description (if one exists).

**Comment by cdjackson** [comment]
> Does this happen on an initial Thing creation or upon a configuration change? As you say "the thing handler throws an exception", I assume that it is in the initialize().

It happens in initialize, but only after an invalid configuration change...

>>we can't update the config

>Why can't we? The updated configuration should be received by the ThingHandler and thus it should have a chance to recover, I would think?

I think the sequence is as follows -:
- Thing starts ok (first time through initialize is ok as the parameter is null)
- UI edits the config, but if it sends back a configuration that doesn't include some parameters, then I believe that Jetty adds them to the DTO as Strings
- The config update gets called - if we don't fix this here, then we're in trouble. In my case, I don't do anything with the parameter in question as it's only read in initialize
- The thing gets disposed/initialized
- During the initialize, we read the parameter, assume it's a BigInt, but it's not as it's been saved as a String. This causes the handler to crash and we're left with a thing with no handler
- A string with no handler can't be updated (or correct me if I'm wrong, but I can't see how)

> Note that the BaseThingHandler also already does a validation of configuration updates 

So maybe the fix is to ensure we always call super?

**Comment by kaikreuzer** [comment]
> The config update gets called

Potentially, this can be called from "anywhere", so the implementation behind it must be robust enough to swallow anything that comes.

> This causes the handler to crash and we're left with a thing with no handler

This definitely must never happen. Do you have a stack trace? I guess we at least need to introduce a try-catch block in `BaseThingHandler.handleConfigurationUpdate`. But I still do not understand how the handler instance can crash? The thread that gets the exception is the one from Jetty handling the REST call...?

> So maybe the fix is to ensure we always call super?

If you override `handleConfigurationUpdate` and you don't do any proper validation yourself, you definitely should, yes.

**Comment by cdjackson** [comment]
```
09:54:18.726 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while calling thing handler factory 'org.openhab.binding.zwave.internal.ZWaveHandlerFactory@a407fd': java.lang.String cannot be cast to java.math.BigDecimal
java.lang.ClassCastException: java.lang.String cannot be cast to java.math.BigDecimal
        at org.openhab.binding.zwave.handler.ZWaveThingHandler.initialize(ZWaveThingHandler.java:107)[235:org.openhab.binding.zwave:2.0.0.201602222325]
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:130)[94:org.eclipse.smarthome.core.thing:0.8.0.201601201718]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:507)[94:org.eclipse.smarthome.core.thing:0.8.0.201601201718]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:1)[94:org.eclipse.smarthome.core.thing:0.8.0.201601201718]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_60]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_60]
09:54:18.727 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:fibaro_fgk101_00_000:152983ff067:node10' changed from INITIALIZING to UNINITIALIZED (HANDLER_INITIALIZING_ERROR): java.lang.String cannot be cast to java.math.BigDecimal
```

**Comment by kaikreuzer** [comment]
Ok, so it is on the initial creation after all then and not in the `handleConfigurationUpdate`.
Nonetheless, the framework should try to reinitialize the handler on an update - I will have to analyze with @sbussweiler why this isn't working. Do you potentially have another stacktrace of what exactly happens if you now try to replace the config by something valid?

**Comment by kaikreuzer** [comment]
I just noticed the version `0.8.0.201601201718` in your stacktrace. The PR https://github.com/eclipse/smarthome/pull/769 has been merged 28 day ago, so it is not yet in the code you use. A lot of this handling has changed there meanwhile - so I would suggest to test the same against the latest codebase before we further analyse. Btw, a new stable ESH build is just finishing (https://hudson.eclipse.org/smarthome/job/SmartHomeDistribution-Stable/34/), so it will be part of tonights openHAB build.
